types/react/index.d.ts

```diff
@@ -255,8 +255,30 @@ declare namespace React {
         unstable_observedBits?: number;
     }
 
-    type Provider<T> = ComponentType<ProviderProps<T>>;
-    type Consumer<T> = ComponentType<ConsumerProps<T>>;
+    // TODO: similar to how Fragment is actually a symbol, the values returned from createContext,
+    // forwardRef and memo are actually objects that are treated specially by the renderer; see:
+    // https://github.com/facebook/react/blob/v16.6.0/packages/react/src/ReactContext.js#L35-L48
+    // https://github.com/facebook/react/blob/v16.6.0/packages/react/src/forwardRef.js#L42-L45
+    // https://github.com/facebook/react/blob/v16.6.0/packages/react/src/memo.js#L27-L31
+    // However, we have no way of telling the JSX parser that it's a JSX element type or its props other than
+    // by pretending to be a normal component.
+    //
+    // We don't just use ComponentType or SFC types because you are not supposed to attach statics to this
+    // object, but rather to the original function.
+    interface ExoticComponent<P = {}> {
+        /**
+         * # WARNING
+         *
+         * Exotic components are not actually callable
+         *
+         * @deprecated Exotic components are not callable
+         */
+        (props: P): (ReactElement<any>|null);
```

Comment:
Aside: I'm surprised that this actually works with how the compiler resolves JSX value-based elements. I had thought that the return type of this call signature needs to be assignable to `JSX.ElementClass`, but I guess if the tests are passing, then it still works?

Reply:
This looks like an SFC signature to the compiler, so it's return value is only checked against `JSX.Element | null`.